### PR TITLE
[client] Undo PySpark's monkey patching of namedtuples for PickleStub

### LINKED ISSUE #19032: [Ray Client][Bug] Ray conflicts with PySpark on namedtuple customization
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### What happened + What you expected to happen

Our customer has code depends on pyspark, which monkey patches the namedtuple class's serialization.  When Ray client serialize a ClientObjectRef, it will try to create a PickleStub which extendes namedtuple class.  While pyspark serialized it, Ray is not able to recognize it, and causing failure.
```
    return ray.get(driver_func.remote(*args, **kwargs))
  File "/usr/lib/python3.6/site-packages/ray/util/client/common.py", line 345, in remote
    return return_refs(ray.call_remote(self, *args, **kwargs))
  File "/usr/lib/python3.6/site-packages/ray/util/client/api.py", line 102, in call_remote
    return self.worker.call_remote(instance, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/ray/util/client/worker.py", line 460, in call_remote
    return self._call_schedule_for_task(task, instance._num_returns())
  File "/usr/lib/python3.6/site-packages/ray/util/client/worker.py", line 477, in _call_schedule_for_task
    raise cloudpickle.loads(ticket.error)
AssertionError
```

The workaround is adding a bypass in the pyspark serializers file like below, in our driver node, make this manual change: in pyspark/serializers.py
```
def _hack_namedtuple(cls):
    """ Make class generated by namedtuple picklable """
    name = cls.__name__
    fields = cls._fields

    def __reduce__(self):
        # Ray PickleStub will use its original pickler, do not hack other namedtuple subclasses
        if 'PickleStub(' in self.__doc__:
            return super(cls, self).__reduce__()
        return (_restore, (name, fields, tuple(self)))
    cls.__reduce__ = __reduce__
    cls._is_namedtuple_ = True
    return cls 
```

### Reproduction script

Repro like below:
```
ray start --head
```
ray client connect to the ray
in python run code below:
```
from pyspark.sql import DataFrame

from typing import Any, List
import ray
import os
import ray.util
ray_client_connection_str = 'str.to.cluster'
connection_output = ray.util.connect(ray_client_connection_str)

import time
from pyarrow import fs
train_dir = '/path/to/train_data'
normal nodes.
batch_size = 100
batch_size = batch_size
num_workers = 2
ray_pipeline = ray.data.read_parquet(train_dir)

total_rows = ray_pipeline.count()
split_idx = total_rows // num_workers

# this will fail with cloudpickle error
splits = ray_pipeline.split_at_indices([split_idx * i + 10 for i in range(1, num_workers)])
```
remove the first from pyspark.sal import ...  will not trigger the error and splits will work just fine.

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

PySpark hijacks the namedtuple class, patching it for serializability:

https://spark-test.github.io/pyspark-coverage-site/pyspark_serializers_py.html#n373

However, this leads to incompatibilities with cloudpickle:

https://issues.apache.org/jira/browse/SPARK-22674

For Ray, this leads to problems with the Ray client when PySpark is imported:

https://github.com/ray-project/ray/issues/19032

This PR fixes the `PickleStub` serializability by reverting the patch for this specific named tuple.

## Related issue number

closes #19032

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
Is it possible to have a test to verify pyspark works with Ray client (and maybe in Ray cluster too)?

**Comment by krfricke** [comment]
The test is passing, can you take another look @mwtian @scv119 @ericl? Thanks

File: python/ray/tests/test_client_compat.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import pytest
```
Comment (ericl) [inline python/ray/tests/test_client_compat.py:1]: This test is failing on Windows.

**Review by ericl — COMMENTED** [review]
LGTM but please remember to check the Windows build!!!

**Comment by krfricke** [comment]
Oh, sorry, I missed that! Skipping the test on windows for now.

**Review by ericl — APPROVED** [review]

**Pull request merged**